### PR TITLE
OCS routes to appframework

### PR DESCRIPTION
**Pull request by rullzer**
Finally add AppFramework capabilities to our OCS endpoints

Basic overview:
- added `ocs` part in route array
- add new route collection for apps. `<APP>.ocs`
- route is resolved at `/ocsapp/...` (To keep the old OCS routes working... all fine since we do this internally)

TODO:
- [x] Fix tests
- [x] More polishing
- [x] Tests!

@DeepDiver1975 FYI

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
conflicting

**Comment by DeepDiver1975** [comment]
@cmonteroluque @PVince81 I'm setting this to 9.2

**Comment by cmonteroluque** [comment]
yes, thanks

File: apps/files_sharing/lib/controllers/foocontroller.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Http\OCSResponse;
+
+/**
+ * Class ExternalSharesController
+ *
+ * @package OCA\Files_Sharing\Controllers
+ */
+class FooController extends Controller {
```
Comment (PVince81) [inline apps/files_sharing/lib/controllers/foocontroller.php]: FooController ? Explain in the description why is it called so. Also, the PHPDoc says another name

Comment (DeepDiver1975) [inline apps/files_sharing/lib/controllers/foocontroller.php]: It's still in here? pfff .... :kill_it_with_fire:

File: lib/public/AppFramework/OCSController.php [hunk 2]
```diff
@@ -100,4 +101,20 @@ private function buildOCSResponse($format, $data) {
...
 	}
 
+	/**
+	 * Serializes and formats a response
+	 * @param mixed $response the value that was returned from a controller and
+	 * is not a Response instance
+	 * @param string $format the format for which a formatter has been registered
+	 * @throws \DomainException if format does not match a registered formatter
+	 * @return Response
+	 * @since 7.0.0
```
Comment (DeepDiver1975) [inline lib/public/AppFramework/OCSController.php]: yes - the method exists in the base class since 7.0

File: ocs/v1.php [hunk 3]
```diff
@@ -53,13 +59,33 @@
 	\OC::$server->getL10NFactory()->setLanguageFromRequest();
 
 	OC::$server->getRouter()->match('/ocs'.\OC::$server->getRequest()->getRawPathInfo());
+	return;
+} catch (ResourceNotFoundException $e) {
+	// Fall true the not found
```
Comment (PVince81) [inline ocs/v1.php]: OC_Response 404 ?

Comment (DeepDiver1975) [inline ocs/v1.php]: Should be named 'Fall through ...' :see_no_evil: - and be taken care of in the try-catch below

File: ocs/v1.php [hunk 4]
```diff
@@ -53,13 +59,33 @@
...
+
+/*
+ * Then we try the old OCS routes
+ */
+try {
+	if(!\OC::$server->getUserSession()->isLoggedIn()) {
+		OC::handleLogin(\OC::$server->getRequest());
+	}
+	OC::$server->getRouter()->match('/ocsapp'.\OC::$server->getRequest()->getRawPathInfo());
+} catch (LoginException $e) {
```
Comment (PVince81) [inline ocs/v1.php]: does the first router also throw LoginException ? might need to add this there too ?

**Comment by PVince81** [comment]
Code looks good so far with a few minor issues.

File: ocs/v1.php (hunk 4 above)
Comment (DeepDiver1975) [inline ocs/v1.php]: not as far as I'm aware of - LoginException is throw out of OC::handleLogin(\OC::$server->getRequest()) which is not in the first try-catch

**Review by DeepDiver1975 — COMMENTED** [review]
> assuming that a leading slash guaranteed here

it is afaik

**Comment by PVince81** [comment]
@DeepDiver1975 not sure whether I understand trusted servers, but I observed a difference in access_log between master and this branch.

To reproduce:
1. Setup a second OC server OC_B, for example 9.0
1. In OC_B, add OC_A as trusted server (OC_A being the instance from this PR or master)
1. Then run cron.php for OC_B

On master, I see these:
```
127.0.0.1 - - [11/Oct/2016:15:16:26 +0200] "POST /owncloud/ocs/v2.php/apps/federation/api/v1/request-shared-secret?format=json HTTP/1.1" 403 79 "-" "ownCloud Server Crawler"
```
(why 403 ?? No idea)

On this branch, it's 401:
```
127.0.0.1 - - [11/Oct/2016:15:17:16 +0200] "POST /owncloud/ocs/v2.php/apps/federation/api/v1/request-shared-secret?format=json HTTP/1.1" 401 43 "-" "ownCloud Server Crawler"
```

Could be a bug ?

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 not sure whether I understand trusted servers, but I observed a difference in access_log between master and this branch.

cries for integration tests ....

**Comment by PVince81** [comment]
- [ ] TEST: retest "sharees" route

I'll stop testing now, the base URL's missing slash breaks everything.

**Comment by DeepDiver1975** [comment]
> I'll stop testing now, the base URL's missing slash breaks everything.

hmmm .... what's your setup? OC in a subfolder? THX

**Comment by PVince81** [comment]
```
$CONFIG = array (
  'instanceid' => 'ockbfur1449k',
  'passwordsalt' => '8XwiihLzH33ylvI5/mkaTQsX6xUvuO',
  'secret' => 'xbVTAAt+cyzlI+qs5fxf4+xpAbqDn3aNN2B5eJ7bLpQ+bi7q',
  'trusted_domains' => 
  array (
    0 => 'localhost',
  ),
  'datadirectory' => '/srv/www/htdocs/owncloud/data',
  'overwrite.cli.url' => 'http://localhost',
  'dbtype' => 'mysql',
  'version' => '9.2.0.2',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'owncloud',
  'dbpassword' => 'owncloud42',
  'logtimezone' => 'UTC',
  'installed' => true,
  'has_internet_connection' => false,
  'appstoreenabled' => false,
  'appstore.experimental.enabled' => true,
  'loglevel' => 0,
  'apps_paths' => 
  array (
    0 => 
    array (
      'path' => '/srv/www/htdocs/owncloud/apps',
      'url' => '/apps',
      'writable' => true,
    ),
    1 => 
    array (
      'path' => '/srv/www/htdocs/owncloud/apps2',
      'url' => '/apps2',
      'writable' => false,
    ),
    2 => 
    array (
      'path' => '/srv/www/htdocs/owncloud/apps3',
      'url' => '/apps3',
      'writable' => false,
    ),
  ),
);
```
this is from my dev env.

URL is "http://localhost/owncloud", so subfolder, yes.

Would it be possible to not touch that part of code at all to avoid further regressions ?

**Comment by DeepDiver1975** [comment]
> Would it be possible to not touch that part of code at all to avoid further regressions ?

I don't see one - there is a cyclic dependency otherwise ...

**Comment by PVince81** [comment]
Ok, now I don't see discrepancies any more. Looks good :+1:

**Comment by SergioBertolinSG** [comment]
This PR has introduced a regression in the ocs v2 answer : https://github.com/owncloud/core/pull/27076

**Pull request merged**